packages/@react-spectrum/tag/src/Tag.tsx

```diff
@@ -10,82 +10,105 @@
-
-  let removable = isGroupRemovable !== undefined ? isGroupRemovable : isRemovable;
-  let disabled = isGroupDisabled !== undefined ? isGroupDisabled : isDisabled;
-  let isInvalid = (validationState !== undefined ? validationState : groupValidationState) === 'invalid';
+  let {isFocusVisible, focusProps} = useFocusRing({within: true});
+  let tagRef = useRef();
+  let labelRef = useRef();
   let {clearButtonProps, labelProps, tagProps} = useTag({
     ...props,
-    isRemovable: removable,
-    isDisabled: disabled,
-    validationState: validationState !== undefined ? validationState : groupValidationState,
-    onRemove: props.onRemove || onRemove,
-    role
-  });
-  let {role: buttonRole, ...otherButtonProps} = clearButtonProps;
-  let icon = props.icon || (isInvalid && <Alert />);
+    isRemovable,
+    isDisabled,
+    item,
+    onRemove,
+    tagRef,
+    labelRef
+  }, state);
 
   return (
-    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
-      <div
-        {...styleProps}
-        {...mergeProps(tagProps, hoverProps)}
-        className={classNames(
-          styles,
-          'spectrum-Tags-item',
-          {
-            'is-disabled': disabled,
-            // 'is-selected': isSelected,
-            'spectrum-Tags-item--removable': removable,
-            'is-invalid': isInvalid,
-            'is-hovered': isHovered
-          },
-          styleProps.className
-        )}>
-        {icon && React.cloneElement(icon, {size: 'S', UNSAFE_className: classNames(styles, 'spectrum-Tags-itemIcon')})}
-        <span
-          {...labelProps}
-          className={classNames(styles, 'spectrum-Tags-itemLabel')}>
-          {props.children}
-        </span>
-        {removable &&
-          <span role={buttonRole}>
-            <ClearButton
-              tabIndex={tagProps.tabIndex}
-              focusClassName={classNames(styles, 'is-focused')}
-              UNSAFE_className={classNames(styles, 'spectrum-Tags-itemClearButton')}
-              {...otherButtonProps} />
-          </span>
+    <div
+      {...mergeProps(tagProps, hoverProps, focusProps, otherProps)}
+      role={tagProps.role}
+      className={classNames(
+        styles,
+        'spectrum-Tags-item',
+        {
+          'is-disabled': isDisabled,
+          'focus-ring': isFocusVisible,
+          'is-focused': isFocusVisible,
+          'not-removable': !isRemovable,
+          'is-hovered': isHovered
         }
-      </div>
-    </FocusRing>
+      )}>
+      <SlotProvider
+        slots={{
+          tagLabel: {UNSAFE_className: classNames(styles, 'spectrum-Tag-label')}
+        }}>
+        <TagLabel {...mergeProps(props, labelProps)} isRemovable={isRemovable} />
+        {isRemovable && <TagRemoveButton item={item} {...clearButtonProps} UNSAFE_className={classNames(styles, 'spectrum-Tag-action')} />}
+      </SlotProvider>
+    </div>
+  );
+}
+
+function TagLabel(props) {
+  let {role, children, isRemovable} = props;
```

Comment:
Should the `id` from `labelProps` be extracted and applied on the span? What about the props from `useGridCell` that also within `labelProps`?